jupyterhub/proxy.py

```diff
@@ -55,14 +65,32 @@ def stop(self):
 
         Will be called during teardown if should_start is True.
         """
+    
+    def validate_routespec(self, routespec):
+        """Validate a routespec
+        
+        - Checks host value vs host-based routing.
+        - Ensures trailing slash on path.
+        """
+        # check host routing
+        host_route = not routespec.startswith('/')
+        if host_route and not self.host_routing:
+            raise ValueError("Cannot add host-based route %r, not using host-routing" % routespec)
+        if self.host_routing and not host_route:
+            raise ValueError("Cannot add route without host %r, using host-routing" % routespec)
+        # add trailing slash
+        if not routespec.endswith('/'):
+            return routespec + '/'
+        else:
+            return routespec
 
     @gen.coroutine
     def add_route(self, routespec, target, data):
         """Add a route to the proxy.
 
         Args:
-            routespec (str): A specification for which this route will be matched.
-                Could be either a url_prefix or a fqdn.
+            routespec (str): A URI (excluding scheme) for which this route will be matched,
```

Comment:
It's technically a URL prefix we're matching, rather than a URI. We're skipping query params and fragments too, for instance.